BLD: Use the new hypotl on Cygwin, rather than defaulting to npy_hypot

DWesl opened the pull request:
Cygwin has a new `hypotl` that avoids overflow for large numbers.  The new function was released in 3.2 or 3.3 a few weeks ago.  NumPy currently uses its fallback/reference implementation, which is based on `npy_hypot`.

I don't know if there should be a delay for a bit so most people can update their installation.
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

DWesl committed:
BLD: Use the new hypotl on Cygwin.

Available since 3.2 or 3.3, I think.

seberg commented:
LGTM.  Just not quite up-to-date here:  we don't worry about version checks for the cygwin blocklist?

DWesl committed:
BLD: Re-enable the function overrides on old Cygwin.

I'm pretty sure CYGWIN_DLL_VERSION_MAJOR == 3003 because the Cygwin
release version is 3.3.x.  We'll see in a bit.

DWesl commented:
I wasn't sure how to get the Cygwin version from the C preprocessor, but this encouraged me to go find them.  This should now use NumPy's versions of `hypotl` and `cabsl` for Cygwin < 3.3, and also Cygwin's versions of `sqrtl` and `modfl` for Cygwin < 3.2.

DWesl committed:
STY: Wrap long lines.

charris commented:
Thanks @DWesl .
